### ISSUE TITLE
[apex] Inconsistency in "Validate CRUD permission before SOQL/DML operation" [ApexCRUDViolation] rule related to SOQL queries on custom metadata Objects.

### ISSUE THREAD

**Issue by Syed-Salesforce** [pmd/pmd#4760]
**Affects PMD Version:** 

**Rule:** : [ApexCRUDViolation](https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexcrudviolation)

Rule Name: Validate CRUD permission before SOQL/DML operation
Rule Key: ApexCRUDViolation

**Description:**

We have performed a Code Scan analysis on the Custom Metadata of the sandbox org.
But there is some inconsistency in flagging out the rule 'Validate CRUD permission before SOQL/DML operation' on one of the class.

what should be the solution for custom metadata object SOQL queries based on the below link?

CRUD permission/field accessibility check is not applicable for custom metadata types.

https://salesforce.stackexchange.com/questions/368961/where-to-set-metadata-type-field-security

https://salesforce.stackexchange.com/questions/275576/do-we-need-to-check-crud-permissions-on-custom-metadata-type-custom-setting-to-p

**Code Sample demonstrating the issue:**

**Rule is flagging out for below type of query:**

```java
try {
    callout = [SELECT first_c FROM first__mdt
               WHERE DeveloperName = :customMetadataName];
} catch
```

**Not flagging out for below type of query:**

```java
List<Second__mdt> secondList = new List<Second_mdt>(
    [Select second_c from second__mdt
     WHERE DeveloperName = :'customMetadataName' LIMIT 1])

```

**Expected outcome:**

Rule should be flagging out in both the above scenarios

PMD should report a violation at line ..., but doesn't. This is a false-negative.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]* : Ant

**Comment by Syed-Salesforce** [comment]
Can someone please check and confirm on above request?

**Comment by adangel** [comment]
I think, that's indeed a false negative. For variable declarations, the rule currently only checks immediate children - but for `secondList` the SoqlExpression is deeper down the tree.

https://github.com/pmd/pmd/blob/e4f2307b061ac440dd1cf1adc5f73a7b62fc99d2/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexCRUDViolationRule.java#L345-L348

Maybe that should be something along the lines:

```java
node.descendants(ASTSoqlExpression.class).forEach(soql -> checkForAccessibility(soql, data));
```

**Comment by adangel** [comment]
This is still failing under PMD 7.
